demo_nodes_cpp/src/topics/rate_control_subscriber.cpp

```diff
@@ -0,0 +1,79 @@
+// Copyright 2024 Sony Group Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <array>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp_components/register_node_macro.hpp"
+#include "rcpputils/join.hpp"
+
+#include "std_msgs/msg/int64.hpp"
+
+#include "demo_nodes_cpp/visibility_control.h"
+
+namespace demo_nodes_cpp
+{
+// Rate-Control const parameters, 0 division remainder, 5 divisor (slower rate)
+constexpr std::array<int64_t, 2> RATE_CONTROL_CONFIG {0, 5};
+
+// Create a RateControlSubscriber class that subclasses the generic rclcpp::Node base class.
+// The main function below will instantiate the class as a ROS node.
+class RateControlSubscriber : public rclcpp::Node
+{
+public:
+  DEMO_NODES_CPP_PUBLIC
+  explicit RateControlSubscriber(const rclcpp::NodeOptions & options)
+  : Node("rate_control_subscriber", options)
+  {
+    // Create a callback function for when messages are received.
+    auto callback =
+      [this](const std_msgs::msg::Int64 & msg) -> void
+      {
+        RCLCPP_INFO(
+          this->get_logger(),
+          "I receive an message data: [%ld]", msg.data);
+      };
+
+    // Initialize a subscription with a content filter to receive data with
+    // one-fifth 1/5 frequency on the subscription. (2 Hz but publisher 10 Hz)
+    rclcpp::SubscriptionOptions sub_options;
+    sub_options.content_filter_options.filter_expression = "%0 == data % %1";
+    sub_options.content_filter_options.expression_parameters = {
+      std::to_string(RATE_CONTROL_CONFIG[0]),
+      std::to_string(RATE_CONTROL_CONFIG[1])
+    };
```

Comment:
This cannot be compiled with either fast-dds or connextdds, since DDS Content Filtered Topic expressions generally don't support mathematical operations directly within the filter itself.

```bash
root@tomoyafujita:~/ros2_ws/colcon_ws# ros2 run demo_nodes_cpp rate_control_subscriber                                       2024-06-10 21:31:15.492 [DDSSQLFILTER Error] PARSE ERROR: :1:0(0): parse error matching eprosima::fastdds::dds::DDSSQLFilter::FilterExpression
%0 == data % %1
^ -> Function parse_filter_expression
2024-06-10 21:31:15.492 [PARTICIPANT Error] Could not create filter of class DDSSQL for expression "%0 == data % %1 -> Function create_contentfilteredtopic
```

```bash
[D0000|CREATE CFTopic|T=rt/counter_ContentFilterTopic9] DDS_SqlFiltererror:syntax error, unexpected "="
[D0000|CREATE CFTopic|T=rt/counter_ContentFilterTopic9] DDS_SqlFilter_compileWithOptimizationLevel:SQL compiler failed with error-code: -1 (Syntax error)
[D0000|CREATE CFTopic|T=rt/counter_ContentFilterTopic9] PRESParticipant_createContentFilteredTopic:content filter compile error 1
[D0000|CREATE CFTopic|T=rt/counter_ContentFilterTopic9] DDS_ContentFilteredTopic_createI:!create DDS_ContentFilteredTopic
[D0000|CREATE CFTopic|T=rt/counter_ContentFilterTopic9] DDS_DomainParticipant_create_contentfilteredtopic_with_filter:!create content filtered topic
[ERROR] [1718080313.762681694] [rmw_connextdds]: failed to create content-filtered topic: name=rt/counter_ContentFilterTopic9, filter=%0 == data % %1
[ERROR] [1718080313.762808619] [rmw_connextdds]: failed to allocate RMW_Connext_Subscriber
[ERROR] [1718080313.762815767] [rmw_connextdds]: failed to create RMW subscription
```